<a href="https://colab.research.google.com/github/Sheik-info/20262027-UE-DataMining-Project/blob/main/Base_notebook_to_complete.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# UE Data Mining - Projet

Notebook à compléter dans le cadre du projet de l’UE Data Mining des Master 2 Intelligence Artificielle et Master 2 Data Science de l’année scolaire 2026-2027.

* Nom du groupe :
* Membres du groupe :
    * Prénom NOM
    * Ghilas Leticia


## Pensez à mettre des titres

Et du texte pour expliquer les différentes étapes de votre raisonnement, les résultats obtenus et les conclusions que vous en avez tiré.

Pour rappel, ce notebook sera aussi votre rapport sur ce projet, donc soyez précis et rigoureux dans ce que vous décrivez et écrivez.

In [2]:
# TODO
!git clone https://github.com/Sheik-info/20262027-UE-DataMining-Project.git
%cd 20262027-UE-DataMining-Project

Cloning into '20262027-UE-DataMining-Project'...
remote: Enumerating objects: 123, done.
remote: Counting objects: 100% (59/59), done.
remote: Compressing objects: 100% (48/48), done.
remote: Total 123 (delta 35), reused 14 (delta 9), pack-reused 64 (from 1)
Receiving objects: 100% (123/123), 1.73 MiB | 13.76 MiB/s, done.
Resolving deltas: 100% (44/44), done.
/content/20262027-UE-DataMining-Project


In [3]:
!git pull

Already up to date.


**Ecology**

**Health**

# 1. Chargement et première exploration du jeu de données

Nous commençons par charger le fichier `cardiac_recordings.csv`, situé dans le dossier `data/health/` du projet.

L’objectif de cette première étape est de comprendre la **structure générale du jeu de données** avant toute opération de nettoyage :

- le nombre d’observations ;
- le nombre de variables ;
- le contenu des colonnes ;
- les types de données.

In [ ]:
import pandas as pd

df = pd.read_csv("data/health/cardiac_recordings.csv")

Nous vérifions ensuite les **dimensions du jeu de données** afin de connaître :

- le nombre de lignes ;
- le nombre de colonnes.

In [ ]:
df.shape

(44, 19)

Le jeu de données contient 44 observations et 19 variables.

### Aperçu des premières observations

Nous affichons les premières lignes du jeu de données afin d’obtenir un premier aperçu de son contenu et de comprendre ce que représente une observation.

In [ ]:
df.head()

,recording_id,patient_number,encounter_number,recording_date,recording_context,facility_license_number,reviewing_license_number,device_serial_number,device_model,sampling_rate_hz,sampling_interval_ms,duration_seconds,channel_count,sample_count,amplitude_unit,automated_interpretation,recording_notes_code,device_quality,archive_status
0,CRD-410031,PAT-31162,ENC-000004,2022-06-21,rest,FAC-VLN-101,LIC-VLN-RES-1016,DEV-CARD-07034,Vantage Rhythm Monitor,100,10,12.0,2,1200,normalized_unit,complex,rhythm_variability,80.9,finalized
1,CRD-410019,PAT-31162,ENC-000004,2022-06-21,technical_repeat,FAC-VLN-101,LIC-VLN-RES-1016,DEV-CARD-07044,HelixCard Duo,100,10,30.0,2,3000,normalized_unit,irregular,repeat_recommended,88.0,finalized
2,CRD-410025,PAT-31040,ENC-000007,2017-03-16,rest,FAC-AVL-201,LIC-VLN-MED-1003,DEV-CARD-07027,Asterion CV-3,125,8,12.0,2,1500,normalized_unit,review_required,no_note,84.8,finalized
3,CRD-410016,PAT-31115,ENC-000017,2025-09-11,rest,FAC-AVL-301,LIC-VLN-MED-1006,DEV-CARD-07019,Asterion CV-2,125,8,30.0,2,3750,normalized_unit,inconclusive,signal_overlap,86.7,finalized
4,CRD-410014,PAT-31233,ENC-000033,2023-04-25,rest,FAC-AVL-301,LIC-VLN-MED-1001,DEV-CARD-07022,HelixCard Duo,125,8,18.0,2,2250,normalized_unit,complex,rhythm_variability,93.0,finalized


In [ ]:
%cd /content/20262027-UE-DataMining-Project

/content/20262027-UE-DataMining-Project


In [ ]:
import os

os.listdir("data/health")

['clinical_notes.csv',
 'clinical_encounters.csv',
 'clinical_procedures.csv',
 'patient_administration.csv',
 'cardiac_signal_samples.csv',
 'clinical_observations.csv',
 'cardiac_recordings.csv']

In [ ]:
import pandas as pd
import os

for fichier in os.listdir("data/health"):
    df_temp = pd.read_csv("data/health/" + fichier)

    print("\n==============================")
    print(fichier)
    print("Dimensions :", df_temp.shape)
    print("Colonnes :", list(df_temp.columns))


clinical_notes.csv
Dimensions : (30, 11)
Colonnes : ['document_id', 'patient_number', 'encounter_number', 'document_date', 'document_type', 'author_license_number', 'facility_license_number', 'title', 'text', 'confidentiality_class', 'superseded_document_id']

clinical_encounters.csv
Dimensions : (1069, 12)
Colonnes : ['encounter_number', 'patient_number', 'encounter_start', 'encounter_end', 'facility_license_number', 'encounter_type', 'referral_source', 'attending_license_number', 'service_unit', 'disposition', 'confidentiality_class', 'follow_up_date']

clinical_procedures.csv
Dimensions : (311, 14)
Colonnes : ['procedure_record_id', 'encounter_number', 'patient_number', 'procedure_code', 'procedure_description', 'scheduled_time', 'performed_time', 'performing_facility_license', 'performing_license_number', 'authorization_class', 'material_reference', 'procedure_status', 'outcome_code', 'next_review_date']

patient_administration.csv
Dimensions : (241, 10)
Colonnes : ['patient_numbe

## Nettoyage — `patient_administration.csv`

Ce fichier contient les données administratives des patients (identité, démographie, statut d'enregistrement). C'est la table de référence du jeu de données : tous les autres fichiers (`clinical_encounters`, `clinical_observations`, etc.) y font référence via `patient_number`.

**Problèmes attendus, à vérifier systématiquement :**
1. Doublons de lignes (redondance d'information)
2. Incohérences de format dans les variables catégorielles (casse, orthographe)
3. Plausibilité des valeurs numériques/dates (âges, cohérence chronologique)


In [4]:
# Chargement du fichier patient_administration depuis le repo cloné
import pandas as pd
pd.set_option('display.max_columns', None)

pa = pd.read_csv('data/health/patient_administration.csv')
print(f"Dimensions initiales : {pa.shape}")
pa.head()

Dimensions initiales : (241, 10)


,patient_number,family_name,given_name,date_of_birth,registered_sex,home_region,occupation_text,registration_date,primary_facility_license,record_status
0,PAT-31162,Varen,Eli,1962-04-19,male,Valenne Republic,corporate adviser,2018-01-01,FAC-VLN-101,inactive
1,PAT-31040,Valen,Jonah,1972-06-12,male,Valenne Republic,research professional,2017-02-01,FAC-AVL-201,active
2,PAT-31115,Oren,Nadia,1979-09-03,female,Valenne Republic,clinical specialist,2018-05-01,FAC-AVL-301,active
3,PAT-31186,Kader,Soren,1969-02-11,male,Nordhaven Union,corporate adviser,2017-06-01,FAC-VLN-101,active
4,PAT-31215,Roe,Ilan,1976-07-07,male,Avelora Federation,research professional,2018-05-01,FAC-AVL-201,active


### 1. Détection et suppression des doublons

On vérifie s'il existe des lignes strictement identiques (même patient, mêmes informations répétées), ce qui correspond à une redondance d'information pure (cf. cours 1 — "Gérer les redondances d'informations").

In [5]:
# Identification des lignes dupliquées (toutes colonnes identiques)
doublons = pa[pa.duplicated(keep=False)]
print(f"Nombre de lignes impliquées dans un doublon : {len(doublons)}")
doublons

Nombre de lignes impliquées dans un doublon : 2


,patient_number,family_name,given_name,date_of_birth,registered_sex,home_region,occupation_text,registration_date,primary_facility_license,record_status
173,PAT-31045,Garen,Olia,2000-08-01,female,Nordhaven Union,transport worker,2021-02-01,FAC-VLN-101,active
240,PAT-31045,Garen,Olia,2000-08-01,female,Nordhaven Union,transport worker,2021-02-01,FAC-VLN-101,active


In [6]:
# Suppression du doublon détecté, réinitialisation de l'index
pa = pa.drop_duplicates().reset_index(drop=True)
print(f"Dimensions après suppression des doublons : {pa.shape}")

Dimensions après suppression des doublons : (240, 10)


### 2. Normalisation des variables catégorielles

On inspecte les valeurs uniques des colonnes catégorielles afin de repérer des variantes d'écriture d'une même valeur (ex. différences de casse), qui empêchent un regroupement correct des catégories lors de l'analyse (cf. cours 1 — "Format inconsistant").

In [7]:
# Inspection des valeurs uniques, avant correction
for col in ['registered_sex', 'record_status', 'home_region']:
    print(f"{col} -> {pa[col].unique()}")

registered_sex -> ['male' 'female' 'nonbinary']
record_status -> ['inactive' 'active' 'deceased' 'ACTIVE']
home_region -> ['Valenne Republic' 'Nordhaven Union' 'Avelora Federation'
 'Kintara Republic' 'Tembara Commonwealth' 'vALENNE rEPUBLIC']


Deux colonnes présentent des incohérences de casse :
- `record_status` : `'active'` et `'ACTIVE'` désignent la même valeur
- `home_region` : `'Valenne Republic'` et `'vALENNE rEPUBLIC'` désignent la même valeur

On normalise ces deux colonnes (minuscules pour `record_status`, casse "titre" pour `home_region`), puis on revérifie que chaque valeur n'apparaît plus que sous une seule forme.

In [8]:
# Normalisation de la casse sur les deux colonnes concernées
pa['record_status'] = pa['record_status'].str.lower()
pa['home_region'] = pa['home_region'].str.title()

# Vérification post-correction
print(f"record_status -> {pa['record_status'].unique()}")
print(f"home_region -> {pa['home_region'].unique()}")

record_status -> ['inactive' 'active' 'deceased']
home_region -> ['Valenne Republic' 'Nordhaven Union' 'Avelora Federation'
 'Kintara Republic' 'Tembara Commonwealth']


### 3. Vérification de la plausibilité de l'âge

On calcule l'âge de chaque patient à partir de `date_of_birth`, puis on examine la distribution (moyenne, écart-type, min, max) afin de détecter d'éventuelles valeurs aberrantes (âge négatif, ou supérieur à l'espérance de vie maximale plausible).

In [9]:
# Conversion de la date de naissance et calcul de l'âge
pa['date_of_birth'] = pd.to_datetime(pa['date_of_birth'])
today = pd.Timestamp.now()
pa['age'] = (today - pa['date_of_birth']).dt.days // 365

print(pa['age'].describe())

count    240.000000
mean      55.691667
std       15.988644
min       23.000000
25%       43.750000
50%       55.000000
75%       69.000000
max       91.000000
Name: age, dtype: float64


Les âges s'étalent entre 23 et 91 ans, avec une moyenne de ~56 ans : aucune valeur n'est aberrante. Aucune correction nécessaire sur cette colonne.

In [10]:
# Diagnostic préalable : on vérifie que toutes les dates suivent le même format avant conversion
mask_anomalie = ~pa['registration_date'].str.match(r'^\d{4}-\d{2}-\d{2}$')
print(f"Nombre de valeurs avec un format non-standard : {mask_anomalie.sum()}")
pa.loc[mask_anomalie, 'registration_date']

Nombre de valeurs avec un format non-standard : 2


,registration_date
51,01/09/2019
144,2019/09/01


Deux valeurs ne respectent pas le format `YYYY-MM-DD` majoritaire : `'2019/09/01'` et `'01/09/2019'`. Comme toutes les autres dates de la colonne ont pour jour `01` (jour fixe, probablement un enregistrement standardisé au 1er du mois), on en déduit que ces deux valeurs désignent la même date, le **1er septembre 2019**, simplement écrite avec un séparateur ou un ordre différent.

In [11]:
# Correction des deux formats non-standard identifiés
pa['registration_date'] = pa['registration_date'].replace({
    '2019/09/01': '2019-09-01',
    '01/09/2019': '2019-09-01'
})

# Conversion en datetime, le format étant désormais uniforme
pa['registration_date'] = pd.to_datetime(pa['registration_date'], format='%Y-%m-%d')

# Vérification de cohérence : aucune inscription ne doit précéder la naissance
incoherent = pa[pa['registration_date'] < pa['date_of_birth']]
print(f"Nombre d'incohérences chronologiques détectées : {len(incoherent)}")
incoherent

Nombre d'incohérences chronologiques détectées : 0


,patient_number,family_name,given_name,date_of_birth,registered_sex,home_region,occupation_text,registration_date,primary_facility_license,record_status,age


### Synthèse du nettoyage — `patient_administration.csv`

| Opération | Résultat |
|---|---|
| Doublons supprimés | 1 ligne (241 → 240 lignes) |
| `record_status` normalisé | 3 catégories propres (`active`, `inactive`, `deceased`) |
| `home_region` normalisé | 5 régions propres |
| Âges vérifiés | Plage 23–91 ans, aucune aberration |
| Format de date corrigé | 2 valeurs sur `registration_date` ramenées au format `YYYY-MM-DD` |
| Cohérence chronologique | Aucune incohérence naissance/inscription détectée |

Le fichier nettoyé est sauvegardé pour être réutilisé dans les étapes suivantes du projet.

In [12]:
# Sauvegarde du fichier nettoyé
pa.to_csv('data/health/patient_administration_clean.csv', index=False)

## Nettoyage — `clinical_procedures.csv`

Ce fichier recense les procédures médicales réalisées (type de procédure, horaires, autorisation, matériel utilisé, résultat). Il est lié à `clinical_encounters` via `encounter_number` et à `patient_administration` via `patient_number`.

**Problèmes attendus, à vérifier :**
1. Doublons de lignes
2. Incohérences de casse dans les variables catégorielles
3. Incohérence de formatage dans le texte libre (espaces superflus)
4. Formats de date hétérogènes (`scheduled_time`, `performed_time`)
5. Valeurs manquantes — à distinguer : absence structurelle vs vraie donnée manquante

In [13]:
cp = pd.read_csv('data/health/clinical_procedures.csv')
print(f"Dimensions initiales : {cp.shape}")
cp.head()

Dimensions initiales : (311, 14)


,procedure_record_id,encounter_number,patient_number,procedure_code,procedure_description,scheduled_time,performed_time,performing_facility_license,performing_license_number,authorization_class,material_reference,procedure_status,outcome_code,next_review_date
0,PRC-000001,ENC-001018,PAT-31221,MRP-42,Targeted metabolic resilience administration,2016-02-18T08:30:00,2016-02-18T09:05:00,FAC-VLN-101,LIC-VLN-MED-1002,restricted_specialist,MTL-2016-0002,completed,completed_without_immediate_complication,2016-03-22
1,PRC-000002,ENC-001019,PAT-31221,MRP-42,Targeted metabolic resilience administration,2018-02-22T08:30:00,2018-02-22T09:05:00,FAC-VLN-101,LIC-VLN-MED-1002,restricted_specialist,MTL-2018-0003,completed,completed_without_immediate_complication,2018-04-01
2,PRC-000003,ENC-001020,PAT-31221,MRP-42,Targeted metabolic resilience administration,2020-03-05T08:30:00,2020-03-05T09:05:00,FAC-VLN-101,LIC-VLN-MED-1002,restricted_specialist,MTL-2020-0004,completed,completed_without_immediate_complication,2020-04-17
3,PRC-000004,ENC-001021,PAT-31221,MRP-42,Targeted metabolic resilience administration,2022-02-17T08:30:00,2022-02-17T09:05:00,FAC-VLN-101,LIC-VLN-MED-1002,restricted_specialist,MTL-2022-0005,completed,completed_without_immediate_complication,2022-03-17
4,PRC-000005,ENC-001022,PAT-31221,MRP-42,Targeted metabolic resilience administration,2024-03-01T08:30:00,2024-03-01T09:05:00,FAC-VLN-101,LIC-VLN-MED-1002,restricted_specialist,MTL-2024-0006,completed,completed_without_immediate_complication,2024-04-03


### 1. Doublons

In [14]:
doublons = cp[cp.duplicated(keep=False)]
print(f"Lignes impliquées dans un doublon : {len(doublons)}")
doublons

Lignes impliquées dans un doublon : 2


,procedure_record_id,encounter_number,patient_number,procedure_code,procedure_description,scheduled_time,performed_time,performing_facility_license,performing_license_number,authorization_class,material_reference,procedure_status,outcome_code,next_review_date
262,PRC-000263,ENC-000334,PAT-31107,CARD-IMG,Advanced cardiac imaging,2021-10-31T09:35:00,2021-10-31T09:45:00,FAC-VLN-101,LIC-VLN-RES-1044,standard,NaN,completed,imaging_completed,2022-08-08
310,PRC-000263,ENC-000334,PAT-31107,CARD-IMG,Advanced cardiac imaging,2021-10-31T09:35:00,2021-10-31T09:45:00,FAC-VLN-101,LIC-VLN-RES-1044,standard,NaN,completed,imaging_completed,2022-08-08


In [15]:
cp = cp.drop_duplicates().reset_index(drop=True)
print(f"Dimensions après suppression : {cp.shape}")

Dimensions après suppression : (310, 14)


### 2. Normalisation de la casse — variables catégorielles

`authorization_class` et `outcome_code` contiennent des variantes de casse pour les mêmes catégories (ex. `'standard'` / `'STANDARD'`), qui doivent être unifiées avant toute analyse par groupe.

In [16]:
for col in ['procedure_code', 'authorization_class', 'procedure_status', 'outcome_code']:
    print(f"{col} -> {cp[col].unique()}")

procedure_code -> ['MRP-42' 'AUT-REV' 'MOB-PNL' 'CARD-IMG' 'OCT-TRK' 'AGE-PNL' 'OCC-PNL']
authorization_class -> ['restricted_specialist' 'research_restricted' 'Research_Restricted'
 'confidential' 'standard' 'STANDARD']
procedure_status -> ['completed']
outcome_code -> ['completed_without_immediate_complication' 'research_protocol_completed'
 'RESEARCH_PROTOCOL_COMPLETED' 'review_completed' 'assessment_completed'
 'imaging_completed' 'panel_completed' 'Assessment_Completed']


In [17]:
# Normalisation en minuscules
cp['authorization_class'] = cp['authorization_class'].str.lower()
cp['outcome_code'] = cp['outcome_code'].str.lower()

print(f"authorization_class -> {cp['authorization_class'].unique()}")
print(f"outcome_code -> {cp['outcome_code'].unique()}")

authorization_class -> ['restricted_specialist' 'research_restricted' 'confidential' 'standard']
outcome_code -> ['completed_without_immediate_complication' 'research_protocol_completed'
 'review_completed' 'assessment_completed' 'imaging_completed'
 'panel_completed']


### 3. Nettoyage du texte libre — `procedure_description`

Cette colonne contient, pour un même `procedure_code`, des variantes avec espaces superflus en début/fin et entre les mots (ex. `'  Mobility  and  balance  panel '` au lieu de `'Mobility and balance panel'`). On uniformise avec un nettoyage d'espaces.

pour la suite de l'analyse : `procedure_description` est entièrement déterminé par `procedure_code` (relation 1 à 1) — c'est une redondance d'information au sens du cours.

In [18]:
import re

# Suppression des espaces multiples et des espaces en début/fin de chaîne
cp['procedure_description'] = cp['procedure_description'].str.strip().apply(lambda s: re.sub(r'\s+', ' ', s))

# Vérification : chaque procedure_code ne doit plus avoir qu'une seule description
print(cp.groupby('procedure_code')['procedure_description'].unique())

procedure_code
AGE-PNL                  [Biological-age laboratory panel]
AUT-REV                        [Autonomic function review]
CARD-IMG                        [Advanced cardiac imaging]
MOB-PNL                       [Mobility and balance panel]
MRP-42      [Targeted metabolic resilience administration]
OCC-PNL                   [Occupational health assessment]
OCT-TRK                       [Ocular tracking assessment]
Name: procedure_description, dtype: object


### 4. Formats de date hétérogènes — `scheduled_time` / `performed_time`

On convertit ces deux colonnes en `datetime` et on isole les valeurs que pandas échoue à interpréter automatiquement, signe d'un format non standard.

In [19]:
cp['scheduled_time_dt'] = pd.to_datetime(cp['scheduled_time'], errors='coerce')
cp['performed_time_dt'] = pd.to_datetime(cp['performed_time'], errors='coerce')

anomalies = cp[cp['scheduled_time_dt'].isna() | cp['performed_time_dt'].isna()]
anomalies[['procedure_record_id', 'scheduled_time', 'performed_time']]

,procedure_record_id,scheduled_time,performed_time
53,PRC-000054,2021-06-08T09:35:00,2021/06/08 09:45:00
70,PRC-000071,2021/12/06 12:35:00,2021-12-06T12:45:00
218,PRC-000219,09-Jul-2021 11:35:00,2021-07-09T11:45:00


Trois formats différents cohabitent dans ces colonnes : `YYYY-MM-DDTHH:MM:SS` (format majoritaire, ISO), `YYYY/MM/DD HH:MM:SS` (séparateur slash), et `DD-Mon-YYYY HH:MM:SS` (mois en texte abrégé). On les reconvertit un par un en forçant chaque format reconnu, plutôt que de laisser pandas deviner (ce qui a échoué pour ces 3 lignes précisément).

In [20]:
def parse_mixed_datetime(s):
    for fmt in ('%Y-%m-%dT%H:%M:%S', '%Y/%m/%d %H:%M:%S', '%d-%b-%Y %H:%M:%S'):
        try:
            return pd.to_datetime(s, format=fmt)
        except ValueError:
            continue
    return pd.NaT  # aucun format reconnu

cp['scheduled_time'] = cp['scheduled_time'].apply(parse_mixed_datetime)
cp['performed_time'] = cp['performed_time'].apply(parse_mixed_datetime)

# On supprime les colonnes temporaires de diagnostic
cp = cp.drop(columns=['scheduled_time_dt', 'performed_time_dt'])

# Vérification : plus aucune valeur manquante après reconversion
print(f"Valeurs non converties - scheduled_time : {cp['scheduled_time'].isna().sum()}")
print(f"Valeurs non converties - performed_time : {cp['performed_time'].isna().sum()}")

# Vérification de cohérence logique : la procédure ne peut pas être réalisée avant d'être programmée
incoherent = cp[cp['performed_time'] < cp['scheduled_time']]
print(f"Incohérences performed < scheduled : {len(incoherent)}")

Valeurs non converties - scheduled_time : 0
Valeurs non converties - performed_time : 0
Incohérences performed < scheduled : 0


### 5. Valeurs manquantes — `material_reference` et `next_review_date`

Avant de décider quoi faire des valeurs manquantes, on regarde si elles sont réparties aléatoirement ou si elles suivent un motif, ce qui change complètement l'interprétation (cf. cours 1 — "l'absence d'information peut être une information en soi").

In [21]:
# Répartition des valeurs manquantes de material_reference par type de procédure
print(cp.groupby('procedure_code')['material_reference'].apply(lambda s: f"{s.isna().sum()}/{len(s)} manquants"))

procedure_code
AGE-PNL     39/39 manquants
AUT-REV     43/43 manquants
CARD-IMG    41/41 manquants
MOB-PNL     37/37 manquants
MRP-42       0/50 manquants
OCC-PNL     49/49 manquants
OCT-TRK     51/51 manquants
Name: material_reference, dtype: object


`material_reference` est manquant pour **100% des procédures**, sauf pour `MRP-42` (0% manquant, 50/50 renseigné). Ce n'est donc pas une donnée manquante aléatoire : seule la procédure `MRP-42` (administration d'un produit, probablement une injection) utilise un matériel traçable par référence. Pour les autres codes, l'absence est **structurelle et informative** (« pas de matériel utilisé » plutôt qu'une donnée perdue) — on ne l'impute pas, on la laisse telle quelle.

`next_review_date` est manquant entre 15% et 27% selon le type de procédure, de façon plus diffuse — ça ressemble davantage à « pas de suivi prévu pour ce patient » qu'à une erreur de saisie. Décision à documenter dans le rapport : on conserve ces valeurs manquantes telles quelles (elles seront interprétées comme « aucun suivi programmé » plutôt qu'imputées).

### Synthèse du nettoyage — `clinical_procedures.csv`

| Opération | Résultat |
|---|---|
| Doublons supprimés | 1 ligne (311 → 310 lignes) |
| `authorization_class`, `outcome_code` normalisés | Casse unifiée |
| `procedure_description` nettoyé | Espaces superflus supprimés (redondant avec `procedure_code`, à noter pour l'analyse) |
| `scheduled_time` / `performed_time` reconvertis | 3 formats de date hétérogènes unifiés en datetime, cohérence chronologique vérifiée |
| `material_reference` manquant | Conservé tel quel — absence structurelle (seule `MRP-42` en a) |
| `next_review_date` manquant | Conservé tel quel — interprété comme absence de suivi prévu |

In [22]:
cp.to_csv('data/health/clinical_procedures_clean.csv', index=False)

## Nettoyage et analyse — `clinical_observations.csv`

Ce fichier contient les mesures individuelles prises à chaque rencontre clinique (une ligne = une mesure). On traite ce fichier en deux temps :

1. **Nettoyage**  : doublons, casse, formats de valeurs numériques
2. **Détection de patterns et d'anomalies**  : analyse univariée par IQR/Z-score modifié, puis recherche d'associations entre codes d'interprétation anormaux au sein d'une même rencontre (matrice de co-occurrence)

In [23]:
obs = pd.read_csv('data/health/clinical_observations.csv')
print(f"Dimensions initiales : {obs.shape}")
obs.head()

Dimensions initiales : (17736, 14)


,observation_number,encounter_number,patient_number,observation_time,observation_code,observation_name,numeric_value,text_value,unit,reference_range_text,interpretation_code,device_reference,quality_code,assessor_license_number
0,OBS-0000001,ENC-000001,PAT-31162,2018-02-27T09:27:00,RPULSE,Resting pulse,64.0,NaN,bpm,50–100,within_expected_range,DEV-002,acceptable,LIC-VLN-MED-1001
1,OBS-0000002,ENC-000001,PAT-31162,2018-02-27T09:27:30,APICAL,Apical cardiac rate,88.5,NaN,bpm,50–110,within_expected_range,DEV-002,acceptable,LIC-VLN-MED-1001
2,OBS-0000003,ENC-000001,PAT-31162,2018-02-27T09:28:00,PPULSE,Peripheral pulse rate,75.8,NaN,bpm,50–100,within_expected_range,DEV-002,acceptable,LIC-VLN-MED-1001
3,OBS-0000004,ENC-000001,PAT-31162,2018-02-27T09:28:30,CARDOUT,Estimated cardiac output,5.57,NaN,L/min,3.5–8.0,within_expected_range,DEV-002,acceptable,LIC-VLN-MED-1001
4,OBS-0000005,ENC-000001,PAT-31162,2018-02-27T09:29:00,O2SAT,Peripheral oxygen saturation,97.2,NaN,%,94–100,within_expected_range,DEV-002,acceptable,LIC-VLN-MED-1001


In [25]:
doublons = obs[obs.duplicated(keep=False)]
print(f"Lignes dupliquées : {len(doublons)}")
obs = obs.drop_duplicates().reset_index(drop=True)
print(f"Dimensions après suppression : {obs.shape}")

Lignes dupliquées : 0
Dimensions après suppression : (17734, 14)


### 2. Harmonisation des catégories

`interpretation_code` et `quality_code` mélangent plusieurs casses pour une même catégorie (ex. `review` / `REVIEW`). On harmonise en minuscules.

In [26]:
for col in ['interpretation_code', 'quality_code']:
    print(f"{col} avant -> {sorted(obs[col].unique())}")

obs['interpretation_code'] = obs['interpretation_code'].str.lower()
obs['quality_code'] = obs['quality_code'].str.lower()

for col in ['interpretation_code', 'quality_code']:
    print(f"{col} après -> {sorted(obs[col].unique())}")

interpretation_code avant -> ['RECORDED', 'REVIEW', 'Recorded', 'WITHIN_EXPECTED_RANGE', 'Within_Expected_Range', 'discordant', 'recorded', 'review', 'within_expected_range']
quality_code avant -> ['ACCEPTABLE', 'Acceptable', 'acceptable']
interpretation_code après -> ['discordant', 'recorded', 'review', 'within_expected_range']
quality_code après -> ['acceptable']


### 3. Normalisation de `numeric_value`

Certaines valeurs ne se convertissent pas directement en nombre, pour deux raisons distinctes qu'on traite séparément :
- virgule décimale au lieu d'un point (`'84,9'`)
- unité collée au texte (`'82.2 index'`, `'8.07 L/min'`)

On garde la colonne brute intacte (traçabilité) et on construit une colonne numérique propre.

In [27]:
# Étape a : repérer ce qui ne passe pas en numérique tel quel
valeur_numerique_brute = pd.to_numeric(obs['numeric_value'], errors='coerce')
a_corriger = obs['numeric_value'].notna() & valeur_numerique_brute.isna()
print(f"Valeurs à corriger : {a_corriger.sum()}")

# Étape b : on sépare la partie numérique de l'unité éventuelle, puis on gère la virgule
# Motif : un nombre (avec point OU virgule comme séparateur décimal), suivi ou non d'un texte d'unité
extraction = obs['numeric_value'].str.extract(r'^([-+]?\d+(?:[.,]\d+)?)')
nombre_point = extraction[0].str.replace(',', '.', regex=False)

obs['numeric_value_clean'] = pd.to_numeric(nombre_point, errors='coerce')

# Vérification : combien reste non converti après ce traitement ?
reste = obs['numeric_value'].notna() & obs['numeric_value_clean'].isna()
print(f"Valeurs encore non convertibles : {reste.sum()}")

Valeurs à corriger : 340
Valeurs encore non convertibles : 0


In [28]:
# Contrôle de cohérence : les valeurs corrigées restent-elles dans un ordre de grandeur plausible
# par rapport aux autres mesures du même observation_code ?
controle = obs.loc[a_corriged := a_corriger, ['observation_code', 'numeric_value', 'numeric_value_clean']]
display(controle.groupby('observation_code')['numeric_value_clean'].describe())

,count,mean,std,min,25%,50%,75%,max
observation_code,,,,,,,,
AGECOND,5.0,1.000000,1.224745,0.000,0.00000,1.0000,1.0000,3.000
APICAL,22.0,77.650000,9.440579,62.600,71.22500,75.6000,84.6750,99.400
BALANCE,24.0,72.241667,12.812254,48.600,64.05000,71.8000,81.9500,94.100
BIOAGE,20.0,52.440000,17.241028,14.900,41.80000,53.1500,66.2250,81.100
CAMMATCH,3.0,90.666667,3.690980,86.700,89.00000,91.3000,92.6500,94.000
CARDOUT,11.0,5.810000,1.072502,4.700,4.98500,5.6900,6.6300,8.070
COGNITIVE,12.0,85.575000,7.968703,69.100,82.35000,85.0500,89.6250,98.300
CVRISK,3.0,53.166667,14.227555,37.200,47.50000,57.8000,61.1500,64.500
DIABP,10.0,76.580000,10.510291,56.600,71.32500,76.9500,81.7250,93.200


### 4. Détection d'anomalies univariée

Pour chaque `observation_code`on calcule les bornes IQR et le Z-score modifié (basé sur la médiane et l'ecart absolu median, moins sensible aux valeurs extrêmes que la moyenne/écart-type — cf. cours 3, diapo 14).

In [29]:
import numpy as np

def detecter_outliers(groupe):
    valeurs = groupe.dropna()
    if len(valeurs) < 4:
        return pd.Series(dtype=float)

    # IQR
    q1, q3 = valeurs.quantile(.25), valeurs.quantile(.75)
    iqr = q3 - q1
    masque_iqr = (valeurs < q1 - 1.5 * iqr) | (valeurs > q3 + 1.5 * iqr)

    # Z-score modifié (médiane + écart absolu médian)
    mediane = valeurs.median()
    eam = (valeurs - mediane).abs().median()
    z_modifie = (valeurs - mediane) / eam if eam != 0 else pd.Series(0, index=valeurs.index)
    masque_z = z_modifie.abs() > 3.5  # seuil usuel pour le Z-score modifié

    return pd.DataFrame({
        'valeur': valeurs,
        'outlier_iqr': masque_iqr,
        'z_modifie': z_modifie.round(2),
        'outlier_z_modifie': masque_z
    })

resultats_outliers = {}
for code in obs['observation_code'].unique():
    sous_ensemble = obs[obs['observation_code'] == code]['numeric_value_clean']
    r = detecter_outliers(sous_ensemble)
    if len(r) > 0:
        resultats_outliers[code] = r

# Synthèse : nombre d'outliers détectés par méthode, par code
synthese_outliers = pd.DataFrame({
    code: {
        'n': len(r),
        'outliers_iqr': r['outlier_iqr'].sum(),
        'outliers_z_modifie': r['outlier_z_modifie'].sum()
    }
    for code, r in resultats_outliers.items()
}).T.astype(int)

synthese_outliers.sort_values('outliers_z_modifie', ascending=False)

,n,outliers_iqr,outliers_z_modifie
LATSWAY,934,32,58
THORVIB,933,20,33
CARDOUT,941,20,32
COGNITIVE,948,14,26
RPULSE,1017,12,25
LEFTEYE,240,15,24
PPULSE,1017,9,24
THORPEAK,932,10,23
RIGHTEYE,240,11,23
MOBILITY,936,10,14


### 5. Recherche de patterns — co-occurrence des interprétations anormales (cours 3)

On s'intéresse maintenant à un autre type de pattern : certaines mesures sont marquées `review` (à revoir) ou `discordant` (incohérente avec une autre mesure) plutôt qu'`within_expected_range`. On cherche si ces signalements se **produisent ensemble**, au sein d'une même rencontre, pour certains couples de type de mesure — ce qui serait un pattern au sens du cours 3 (itemsets, association).

**Construction des transactions :** chaque rencontre (`encounter_number`) est une transaction. Chaque `observation_code` est un item. Un item vaut 1 dans une transaction si cette mesure a été signalée `review` ou `discordant` lors de cette rencontre.

In [30]:
# Construction de la matrice transactions x items
obs_flag = obs[obs['interpretation_code'].isin(['review', 'discordant'])]

matrice_items = (
    pd.crosstab(obs_flag['encounter_number'], obs_flag['observation_code'])
    .clip(upper=1)  # une mesure signalée plusieurs fois dans la même rencontre compte comme 1
)

print(f"Nombre de transactions (rencontres avec au moins un signalement) : {len(matrice_items)}")
print(f"Items (codes de mesure jamais signalés 'review'/'discordant') : {list(matrice_items.columns)}")
matrice_items.head()

Nombre de transactions (rencontres avec au moins un signalement) : 249
Items (codes de mesure jamais signalés 'review'/'discordant') : ['APICAL', 'LATSWAY', 'OCULAR', 'THORPEAK', 'THORVIB']


observation_code,APICAL,LATSWAY,OCULAR,THORPEAK,THORVIB
encounter_number,,,,,
ENC-000001,0,1,0,1,1
ENC-000002,1,1,0,1,1
ENC-000003,0,0,1,0,1
ENC-000004,1,1,1,1,1
ENC-000005,0,0,1,0,1


In [31]:
# Matrice de co-occurrence brute (cours 3, diapo 4), puis en pourcentage par ligne
co_occurrence = matrice_items.T.dot(matrice_items)
print("Matrice de co-occurrence (effectifs) :")
display(co_occurrence)

co_occurrence_pct = co_occurrence.div(np.diag(co_occurrence), axis=0) * 100
print("\nMatrice de co-occurrence (% par ligne) :")
display(co_occurrence_pct.round(1))

Matrice de co-occurrence (effectifs) :


observation_code,APICAL,LATSWAY,OCULAR,THORPEAK,THORVIB
observation_code,,,,,
APICAL,109,43,20,55,39
LATSWAY,43,123,17,30,42
OCULAR,20,17,42,9,15
THORPEAK,55,30,9,83,37
THORVIB,39,42,15,37,78



Matrice de co-occurrence (% par ligne) :


observation_code,APICAL,LATSWAY,OCULAR,THORPEAK,THORVIB
observation_code,,,,,
APICAL,100.0,39.4,18.3,50.5,35.8
LATSWAY,35.0,100.0,13.8,24.4,34.1
OCULAR,47.6,40.5,100.0,21.4,35.7
THORPEAK,66.3,36.1,10.8,100.0,44.6
THORVIB,50.0,53.8,19.2,47.4,100.0


### 6. Évaluation des associations — support, confiance, lift (cours 3)

Pour chaque paire de codes, on calcule les mesures d'intérêt vues en cours : support (fréquence conjointe), confiance (P(Y|X)), et lift (force de l'association par rapport au hasard).

In [32]:
from itertools import permutations

n_transactions = len(matrice_items)
support_individuel = matrice_items.mean()  # support de chaque item seul

regles = []
for x, y in permutations(matrice_items.columns, 2):
    support_xy = (matrice_items[x] & matrice_items[y]).mean()
    if support_xy == 0:
        continue
    confiance = support_xy / support_individuel[x]
    lift = confiance / support_individuel[y]
    regles.append({
        'X': x, 'Y': y,
        'support(X->Y)': round(support_xy, 3),
        'confiance(X->Y)': round(confiance, 3),
        'lift(X->Y)': round(lift, 3)
    })

df_regles = pd.DataFrame(regles).sort_values('lift(X->Y)', ascending=False)
df_regles

,X,Y,support(X->Y),confiance(X->Y),lift(X->Y)
2,APICAL,THORPEAK,0.221,0.505,1.514
12,THORPEAK,APICAL,0.221,0.663,1.514
15,THORPEAK,THORVIB,0.149,0.446,1.423
19,THORVIB,THORPEAK,0.149,0.474,1.423
3,APICAL,THORVIB,0.157,0.358,1.142
16,THORVIB,APICAL,0.157,0.500,1.142
11,OCULAR,THORVIB,0.060,0.357,1.140
18,THORVIB,OCULAR,0.060,0.192,1.140
17,THORVIB,LATSWAY,0.169,0.538,1.090
7,LATSWAY,THORVIB,0.169,0.341,1.090


### Synthèse — `clinical_observations.csv`

| Étape | Résultat |
|---|---|
| Doublons supprimés | 2 lignes |
| Casse harmonisée | `interpretation_code`, `quality_code` |
| `numeric_value` normalisé | 340 valeurs corrigées (virgules décimales + unités collées au texte), colonne brute conservée |
| Outliers univariés | Détectés par IQR et Z-score modifié, par `observation_code` (voir tableau) |
| Pattern détecté | Les signalements `review`/`discordant` ne concernent que 5 codes (`APICAL`, `LATSWAY`, `OCULAR`, `THORPEAK`, `THORVIB`), sur 249/1017 rencontres |
| Association la plus forte | `APICAL` ↔ `THORPEAK` ↔ `THORVIB` forment un trio dont les anomalies co-occurrent (lift > 1.1 pour chaque paire) ; `OCULAR` et `LATSWAY` sont plus indépendants de ce trio (lift < 1 avec `THORPEAK`), à nuancer vu leur faible support |



In [ ]:
!git status

On branch main
Your branch is up to date with 'origin/main'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	20262027-UE-DataMining-Project/

nothing added to commit but untracked files present (use "git add" to track)


**Logistics**


On explore ici les jeux de données du dossier `data/logistics/` :
- `erp_facilities.csv` : les entrepôts / sites de l'entreprise
- `erp_business_partners.csv` : les partenaires (clients, fournisseurs, transporteurs...)
- `inventory_month_end.csv` : les stocks en fin de mois par site
- `inventory_adjustments.csv` : les ajustements manuels de stock
- `goods_receipts.csv` : les réceptions de marchandises
- `shipments.csv` : les expéditions
- `purchase_order_lines.csv` : les lignes de commandes d'achat

On commence, comme pour les autres parties, par un premier chargement et une exploration rapide avant de nettoyer quoi que ce soit.

 1. Chargement et première exploration
---




In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
os.listdir("data/logistics")

['erp_business_partners.csv',
 'purchase_order_lines.csv',
 'erp_facilities.csv',
 'shipments.csv',
 'inventory_month_end.csv',
 'inventory_adjustments.csv',
 'goods_receipts.csv']

Avant de charger les fichiers, on regarde rapidement leurs dimensions et leurs colonnes.
Cela permet de vérifier la structure des données.


In [ ]:
# Je regarde rapidement les fichiers avant de commencer
for fichier in os.listdir("data/logistics"):
    df = pd.read_csv("data/logistics/" + fichier)

    print("\n==============================")
    print(fichier)
    print("Dimensions :", df.shape)
    print("Colonnes :", list(df.columns))


erp_business_partners.csv
Dimensions : (80, 11)
Colonnes : ['partner_code', 'partner_name', 'partner_role', 'external_registration_number', 'billing_region', 'shipping_region', 'default_currency', 'payment_terms', 'partner_status', 'risk_class', 'onboarding_date']

purchase_order_lines.csv
Dimensions : (651, 16)
Colonnes : ['purchase_order_number', 'purchase_order_line', 'order_date', 'ordering_facility_code', 'supplier_partner_code', 'item_code', 'item_description', 'item_category', 'ordered_quantity', 'ordered_unit', 'unit_price', 'currency', 'requested_delivery_date', 'authorization_class', 'declared_cost_center', 'order_status']

erp_facilities.csv
Dimensions : (30, 11)
Colonnes : ['facility_code', 'facility_name', 'facility_category', 'operating_partner_code', 'street_address', 'region', 'latitude', 'longitude', 'opening_date', 'facility_status', 'storage_capabilities']

shipments.csv
Dimensions : (452, 16)
Colonnes : ['shipment_number', 'shipment_leg', 'dispatch_time', 'arrival_

Les 7 fichiers sont chargés séparément afin de faciliter leur manipulation et leur analyse.

In [ ]:
df_fac = pd.read_csv("data/logistics/erp_facilities.csv")
df_bp = pd.read_csv("data/logistics/erp_business_partners.csv")
df_invm = pd.read_csv("data/logistics/inventory_month_end.csv")
df_inva = pd.read_csv("data/logistics/inventory_adjustments.csv")
df_rcv = pd.read_csv("data/logistics/goods_receipts.csv")
df_shp = pd.read_csv("data/logistics/shipments.csv")
df_po = pd.read_csv("data/logistics/purchase_order_lines.csv")

In [ ]:
dfs = {
    "facilities": df_fac,
    "business_partners": df_bp,
    "inventory_month_end": df_invm,
    "inventory_adjustments": df_inva,
    "goods_receipts": df_rcv,
    "shipments": df_shp,
    "purchase_order_lines": df_po
}

In [ ]:
df_fac.head()

,facility_code,facility_name,facility_category,operating_partner_code,street_address,region,latitude,longitude,opening_date,facility_status,storage_capabilities
0,FCL-1011,Meridian Valenne Headquarters,office,BP-30078,"Meridian Valenne Headquarters, Valenne Central...",Valenne Central District,47.24,7.58,2006-01-01,active,ambient
1,FCL-1024,Meridian Kintara Biological Aggregation Facility,warehouse,BP-30078,Meridian Kintara Biological Aggregation Facili...,Kintara Karst Cluster,4.74,112.28,1997-01-01,active,ambient;secure_cage;2-8C;frozen
2,FCL-1019,Meridian Avelora Transfer Annex,transfer_hub,BP-30078,"Meridian Avelora Transfer Annex, Eastern Avelo...",Eastern Avelora Agricultural Cluster,7.46,101.91,2003-01-01,active,ambient;secure_cage;2-8C;frozen
3,FCL-1020,Meridian Tembara Coastal Hub,transfer_hub,BP-30078,"Meridian Tembara Coastal Hub, Tembara Coastal ...",Tembara Coastal Cluster,2.20,104.55,2010-01-01,active,ambient;secure_cage;2-8C;frozen
4,FCL-1016,Northstar Tembara Cold-Chain Hub,warehouse,BP-30047,"Northstar Tembara Cold-Chain Hub, Tembara Coas...",Tembara Coastal Cluster,2.23,104.58,2008-01-01,active,ambient;secure_cage;2-8C;frozen


Un aperçu des premières lignes est affiché afin de vérifier la structure et le contenu des données.

In [ ]:
df_po.head()

,purchase_order_number,purchase_order_line,order_date,ordering_facility_code,supplier_partner_code,item_code,item_description,item_category,ordered_quantity,ordered_unit,unit_price,currency,requested_delivery_date,authorization_class,declared_cost_center,order_status
0,PO-2014-00001,1,2014-06-24,FCL-1007,BP-30078,ITM-MED-001,General clinical consumable case,medical_consumables,73,case,157.85,VLN,2014-07-08,restricted,Clinical Programmes,completed
1,PO-2015-00001,1,2015-01-22,FCL-1010,BP-30068,ITM-TRN-001,Regional refrigerated transport service,transport_service,1,service,3482.27,VLN,2015-02-01,ordinary,Clinical Services,completed
2,PO-2015-00002,1,2015-01-29,FCL-1030,BP-30066,ITM-COM-001,Secure communication service,communication_service,1,service,1292.19,KRN,2015-02-26,ordinary,General Operations,completed
3,PO-2015-00003,1,2015-02-10,FCL-1010,BP-30080,ITM-LAB-002,Sample preparation material pack,laboratory_material,3,pack,458.67,VLN,2015-03-02,expedited,Clinical Services,completed
4,PO-2015-00004,1,2015-02-12,FCL-1022,BP-30021,ITM-GRA-001,Programme grant transfer,grant,1,grant,26933.28,KRN,2015-02-21,ordinary,Facilities,completed


Les types des colonnes sont vérifiés afin de repérer les données qui ne sont pas dans le format attendu.


In [ ]:
for nom, df in dfs.items():
    print("\n=======================")
    print(nom)
    print(df.dtypes)


facilities
facility_code              object
facility_name              object
facility_category          object
operating_partner_code     object
street_address             object
region                     object
latitude                  float64
longitude                 float64
opening_date               object
facility_status            object
storage_capabilities       object
dtype: object

business_partners
partner_code                    object
partner_name                    object
partner_role                    object
external_registration_number    object
billing_region                  object
shipping_region                 object
default_currency                object
payment_terms                   object
partner_status                  object
risk_class                      object
onboarding_date                 object
dtype: object

inventory_month_end
facility_code               object
reporting_month             object
item_category               object
opening_stoc

In [ ]:
for nom, df in dfs.items():
    print("\n---", nom, "---")

    for colonne in df.columns:
        nb = df[colonne].isna().sum()
        if nb > 0:
            print(colonne, ":", nb)


--- facilities ---
storage_capabilities : 1

--- business_partners ---
external_registration_number : 14
billing_region : 1

--- inventory_month_end ---

--- inventory_adjustments ---
approved_by_role : 1
related_receipt_number : 117
related_shipment_number : 103

--- goods_receipts ---
inspection_status : 4

--- shipments ---
arrival_time : 17

--- purchase_order_lines ---
declared_cost_center : 3


Certaines colonnes contenant des quantités et des prix sont actuellement au format `object`. Elles devront être converties dans un format numérique avant de poursuivre l'analyse.

 2. Nettoyage

---



#### 2.1 Lignes dupliquées

La présence de lignes dupliquées est vérifiée dans chaque fichier.

In [ ]:
for nom, df in dfs.items():
    print(nom, ":", df.duplicated().sum(), "doublon(s)")

facilities : 0 doublon(s)
business_partners : 0 doublon(s)
inventory_month_end : 2 doublon(s)
inventory_adjustments : 0 doublon(s)
goods_receipts : 2 doublon(s)
shipments : 2 doublon(s)
purchase_order_lines : 1 doublon(s)


##### Suppression des doublons


In [ ]:
df_invm = df_invm.drop_duplicates().reset_index(drop=True)
df_rcv = df_rcv.drop_duplicates().reset_index(drop=True)
df_shp = df_shp.drop_duplicates().reset_index(drop=True)
df_po = df_po.drop_duplicates().reset_index(drop=True)

dfs["inventory_month_end"] = df_invm
dfs["goods_receipts"] = df_rcv
dfs["shipments"] = df_shp
dfs["purchase_order_lines"] = df_po

for nom, df in dfs.items():
    print(nom, ":", df.shape[0], "lignes")

facilities : 30 lignes
business_partners : 80 lignes
inventory_month_end : 500 lignes
inventory_adjustments : 140 lignes
goods_receipts : 610 lignes
shipments : 450 lignes
purchase_order_lines : 650 lignes


##### 2.2 Colonnes constantes

Les colonnes qui contiennent une seule valeur, ou qui sont entièrement vides, sont repérées. Elles n'apportent pas de variation pour l'analyse et peuvent donc être supprimées.

In [ ]:
for nom, df in dfs.items():
    cols_constantes = [c for c in df.columns if df[c].nunique(dropna=False) <= 1]

    if cols_constantes:
        print(nom, ":", cols_constantes)

facilities : ['facility_status']
goods_receipts : ['purchase_order_line']
shipments : ['shipment_leg']
purchase_order_lines : ['purchase_order_line']


Suppression des colonnes qui ne varient pas


In [ ]:
df_fac = df_fac.drop(columns=["facility_status"])

df_rcv = df_rcv.drop(columns=["purchase_order_line"])
df_po = df_po.drop(columns=["purchase_order_line"])

df_shp = df_shp.drop(columns=["shipment_leg"])

dfs["facilities"] = df_fac
dfs["goods_receipts"] = df_rcv
dfs["purchase_order_lines"] = df_po
dfs["shipments"] = df_shp

#### 2.3 Nettoyage des colonnes numériques — inventory_month_end

Les colonnes `opening_stock`, `received_quantity`, `outgoing_quantity`, `documented_waste` et `closing_stock` sont actuellement de type `object` alors qu'elles devraient être numériques. Deux problèmes de format sont en cause :

- certaines valeurs utilisent une virgule comme séparateur décimal (ex : `"5,42"`) au lieu d'un point ;
- certaines valeurs contiennent l'unité collée au nombre (ex : `"14.55 sealed_lot"`). Cette unité est redondante avec la colonne `unit` déjà présente, elle peut donc être retirée sans perte d'information.

Une fonction de nettoyage est appliquée à ces colonnes.

In [ ]:
import re

def nettoyer_colonne_numerique(serie):
    s = serie.astype(str).str.strip()
    s = s.str.replace(',', '.', regex=False)
    s = s.str.extract(r'(-?\d+\.?\d*)')[0]
    return pd.to_numeric(s, errors='coerce')

colonnes_a_nettoyer = ['opening_stock', 'received_quantity', 'outgoing_quantity',
                        'documented_waste', 'closing_stock']

for col in colonnes_a_nettoyer:
    df_invm[col] = nettoyer_colonne_numerique(df_invm[col])

df_invm.dtypes

,0
facility_code,object
reporting_month,object
item_category,object
opening_stock,float64
received_quantity,float64
outgoing_quantity,float64
documented_internal_use,float64
documented_waste,float64
closing_stock,float64
unit,object


La colonne `inventory_status` présente aussi une incohérence de casse (`"closed"` / `"CLOSED"`, `"provisional"` / `"PROVISIONAL"`). Elle est uniformisée en minuscules.

In [ ]:
df_invm['inventory_status'] = df_invm['inventory_status'].str.lower()
df_invm['inventory_status'].value_counts()

,count
inventory_status,
closed,444
provisional,38
revised,18


##### Vérification de cohérence comptable

Le stock de fin de mois (`closing_stock`) devrait théoriquement correspondre à :
`opening_stock + received_quantity − outgoing_quantity − documented_internal_use − documented_waste`.

Un écart entre valeur déclarée et valeur recalculée est ajouté comme nouvelle colonne, afin d'être analysé plus tard plutôt que d'être corrigé arbitrairement.

In [ ]:
dfs["inventory_month_end"] = df_invm

In [ ]:
df_invm.dtypes


,0
facility_code,object
reporting_month,object
item_category,object
opening_stock,float64
received_quantity,float64
outgoing_quantity,float64
documented_internal_use,float64
documented_waste,float64
closing_stock,float64
unit,object


In [ ]:
df_invm.head()

,facility_code,reporting_month,item_category,opening_stock,received_quantity,outgoing_quantity,documented_internal_use,documented_waste,closing_stock,unit,inventory_status
0,FCL-1024,2018-02-01,biological_sample,14.55,5.46,0.00,0.49,0.22,16.24,sealed_lot,closed
1,FCL-1024,2018-03-01,biological_sample,16.24,5.00,0.00,1.94,0.65,0.55,sealed_lot,closed
2,FCL-1024,2018-05-01,biological_sample,0.55,0.00,0.00,0.00,0.36,0.17,sealed_lot,closed
3,FCL-1024,2019-01-01,biological_sample,0.17,0.00,1.37,0.00,0.02,0.00,sealed_lot,provisional
4,FCL-1024,2019-08-01,biological_sample,0.00,3.12,8.80,0.00,0.08,0.00,sealed_lot,closed


#### 2.4 Nettoyage des colonnes numériques — goods_receipts

La colonne `received_quantity` présente les mêmes problèmes de format que `inventory_month_end` : virgule décimale et unité collée à la valeur (redondante avec `received_unit`). La même fonction de nettoyage est réutilisée.

In [ ]:
df_rcv['received_quantity'] = nettoyer_colonne_numerique(df_rcv['received_quantity'])
df_rcv['received_quantity'].describe()

,received_quantity
count,610.000000
mean,20.591770
std,22.947117
min,0.780000
25%,1.000000
50%,13.000000
75%,29.000000
max,90.000000


La colonne `receipt_status` présente une incohérence de casse (`"posted"` / `"POSTED"`), uniformisée en minuscules.

In [ ]:
df_rcv['receipt_status'] = df_rcv['receipt_status'].str.lower()
df_rcv['receipt_status'].value_counts()

,count
receipt_status,
posted,576
quality_hold,34


4 lignes présentent une valeur manquante pour `inspection_status`, toutes avec un `receipt_status` égal à `"posted"`. Plutôt que de supposer une valeur, ces cas sont rendus explicites avec la catégorie `"non_renseigne"`.

In [ ]:
df_rcv['inspection_status'] = df_rcv['inspection_status'].fillna('non_renseigne')
df_rcv['inspection_status'].value_counts()

,count
inspection_status,
accepted,505
conditional,67
held,34
non_renseigne,4


In [ ]:
dfs["goods_receipts"] = df_rcv

In [ ]:
df_rcv.dtypes

,0
receipt_number,object
purchase_order_number,object
receipt_date,object
receiving_facility_code,object
supplier_partner_code,object
received_quantity,float64
received_unit,object
batch_reference,object
inspection_status,object
storage_location_code,object


In [ ]:
df_rcv.head()

,receipt_number,purchase_order_number,receipt_date,receiving_facility_code,supplier_partner_code,received_quantity,received_unit,batch_reference,inspection_status,storage_location_code,receipt_status
0,RCV-2014-00001,PO-2014-00001,2014-07-08,FCL-1007,BP-30078,73.0,case,B-2014-44545,held,BIN-35,quality_hold
1,RCV-2015-00002,PO-2015-00001,2015-02-02,FCL-1010,BP-30068,0.9,service,B-2015-58495,accepted,BIN-36,posted
2,RCV-2015-00003,PO-2015-00002,2015-02-28,FCL-1030,BP-30066,1.0,service,B-2015-48474,held,BIN-17,quality_hold
3,RCV-2015-00004,PO-2015-00003,2015-03-02,FCL-1010,BP-30080,3.0,pack,B-2015-71913,accepted,BIN-16,posted
4,RCV-2015-00005,PO-2015-00004,2015-02-21,FCL-1022,BP-30021,1.0,grant,B-2015-31154,accepted,BIN-42,posted


#### 2.5 Nettoyage des colonnes numériques — purchase_order_lines

Les colonnes `ordered_quantity` (unité collée) et `unit_price` (virgule décimale) présentent les mêmes problèmes de format que les fichiers précédents.

In [ ]:
df_po['ordered_quantity'] = nettoyer_colonne_numerique(df_po['ordered_quantity'])
df_po['unit_price'] = nettoyer_colonne_numerique(df_po['unit_price'])
df_po[['ordered_quantity', 'unit_price']].describe()

,ordered_quantity,unit_price
count,650.000000,650.000000
mean,20.964615,3844.168754
std,23.136234,5701.539177
min,1.000000,83.810000
25%,1.000000,560.435000
50%,14.000000,1600.670000
75%,30.000000,4243.115000
max,90.000000,28663.740000


La colonne `item_category` présente des doublons de casse (ex : `"monitoring_equipment"` / `"MONITORING_EQUIPMENT"`), uniformisés en minuscules.

In [ ]:
df_po['item_category'] = df_po['item_category'].str.lower()
df_po['item_category'].value_counts()

,count
item_category,
medical_consumables,79
monitoring_equipment,57
cold_storage_equipment,54
agricultural_product,53
protective_equipment,49
laboratory_material,44
biological_sample,42
office_supply,37
transport_service,33


3 lignes présentent une valeur manquante pour `declared_cost_center`, sans motif apparent. Comme pour `inspection_status`, ces cas sont rendus explicites avec la catégorie `"non_renseigne"` plutôt que de supposer une valeur.

In [ ]:
df_po['declared_cost_center'] = df_po['declared_cost_center'].fillna('non_renseigne')
df_po['declared_cost_center'].value_counts()

,count
declared_cost_center,
Field Programmes,98
Clinical Services,93
Research Services,90
Agricultural Distribution,87
General Operations,84
Facilities,81
Cold Chain,70
Clinical Programmes,26
Field Operations,18


**Point de vigilance pour l'analyse** : `unit_price` est exprimé dans 5 devises différentes (VLN, AVL, TMB, NDH, KRN). Toute comparaison  de prix entre commandes nécessitera une standardisation (a faire plus tard)

In [ ]:
dfs["purchase_order_lines"] = df_po

In [ ]:
df_po.head()

,purchase_order_number,order_date,ordering_facility_code,supplier_partner_code,item_code,item_description,item_category,ordered_quantity,ordered_unit,unit_price,currency,requested_delivery_date,authorization_class,declared_cost_center,order_status
0,PO-2014-00001,2014-06-24,FCL-1007,BP-30078,ITM-MED-001,General clinical consumable case,medical_consumables,73,case,157.85,VLN,2014-07-08,restricted,Clinical Programmes,completed
1,PO-2015-00001,2015-01-22,FCL-1010,BP-30068,ITM-TRN-001,Regional refrigerated transport service,transport_service,1,service,3482.27,VLN,2015-02-01,ordinary,Clinical Services,completed
2,PO-2015-00002,2015-01-29,FCL-1030,BP-30066,ITM-COM-001,Secure communication service,communication_service,1,service,1292.19,KRN,2015-02-26,ordinary,General Operations,completed
3,PO-2015-00003,2015-02-10,FCL-1010,BP-30080,ITM-LAB-002,Sample preparation material pack,laboratory_material,3,pack,458.67,VLN,2015-03-02,expedited,Clinical Services,completed
4,PO-2015-00004,2015-02-12,FCL-1022,BP-30021,ITM-GRA-001,Programme grant transfer,grant,1,grant,26933.28,KRN,2015-02-21,ordinary,Facilities,completed


#### 2.6 Nettoyage des colonnes — inventory_adjustments

`adjustment_quantity` présente une unité collée sur une ligne (redondante avec `unit`), retirée avec la même fonction que précédemment.

In [ ]:
df_inva['adjustment_quantity'] = nettoyer_colonne_numerique(df_inva['adjustment_quantity'])
df_inva['adjustment_quantity'].describe()

,adjustment_quantity
count,140.000000
mean,-0.598857
std,3.995308
min,-7.910000
25%,-3.035000
50%,-1.805000
75%,2.005000
max,7.870000


`adjustment_reason` présente des doublons de casse, uniformisés en minuscules.

In [ ]:
df_inva['adjustment_reason'] = df_inva['adjustment_reason'].str.lower()
df_inva['adjustment_reason'].value_counts()

,count
adjustment_reason,
transfer posting correction,24
cycle count correction,23
quality-hold release,21
damaged packaging write-off,16
unit-of-measure review,12
specialist programme issue,11
procedure stock issue,10
late receipt posting,9
controlled-area consumption,9


In [ ]:
df_inva['approved_by_role'] = df_inva['approved_by_role'].fillna('non_renseigne')

**Observation** : `related_receipt_number` et `related_shipment_number` sont laissés tels quels (valeurs manquantes conservées), car l'absence des deux (85 lignes) correspond probablement à des ajustements sans document source lié.

In [ ]:
dfs["inventory_adjustments"] = df_inva

In [ ]:
df_inva.head()

,adjustment_number,facility_code,adjustment_date,item_category,adjustment_quantity,unit,adjustment_reason,approved_by_role,related_receipt_number,related_shipment_number,adjustment_status
0,ADJ-2014-00001,FCL-1007,2014-07-18,medical_consumables,-1.58,case,specialist programme issue,clinical stores manager,NaN,NaN,posted
1,ADJ-2015-00002,FCL-1007,2015-09-10,medical_consumables,-2.70,case,specialist programme issue,specialist programme administrator,NaN,NaN,posted
2,ADJ-2015-00003,FCL-1024,2015-10-30,monitoring_equipment,3.68,kit,late receipt posting,warehouse manager,RCV-2022-00361,NaN,posted
3,ADJ-2016-00004,FCL-1007,2016-01-20,biological_sample,6.52,sealed_lot,late receipt posting,finance controller,NaN,NaN,pending_review
4,ADJ-2016-00005,FCL-1010,2016-02-18,medical_consumables,-2.43,case,procedure stock issue,quality manager,NaN,NaN,posted


#### 2.7 Nettoyage — erp_facilities
 Ce fichier ne présente pas de problème de format numérique ni de casse. Seule la colonne storage_capabilities contient une valeur manquante



In [ ]:
df_fac['storage_capabilities'] = df_fac['storage_capabilities'].fillna('non_renseigne')
df_fac['storage_capabilities'].value_counts()

,count
storage_capabilities,
ambient;secure_cage,9
ambient;secure_cage;2-8C;frozen,8
ambient,6
2-8C;controlled_access,6
non_renseigne,1


mise à jour du dic

In [ ]:
dfs["facilities"] = df_fac
df_fac.head()

,facility_code,facility_name,facility_category,operating_partner_code,street_address,region,latitude,longitude,opening_date,storage_capabilities
0,FCL-1011,Meridian Valenne Headquarters,office,BP-30078,"Meridian Valenne Headquarters, Valenne Central...",Valenne Central District,47.24,7.58,2006-01-01,ambient
1,FCL-1024,Meridian Kintara Biological Aggregation Facility,warehouse,BP-30078,Meridian Kintara Biological Aggregation Facili...,Kintara Karst Cluster,4.74,112.28,1997-01-01,ambient;secure_cage;2-8C;frozen
2,FCL-1019,Meridian Avelora Transfer Annex,transfer_hub,BP-30078,"Meridian Avelora Transfer Annex, Eastern Avelo...",Eastern Avelora Agricultural Cluster,7.46,101.91,2003-01-01,ambient;secure_cage;2-8C;frozen
3,FCL-1020,Meridian Tembara Coastal Hub,transfer_hub,BP-30078,"Meridian Tembara Coastal Hub, Tembara Coastal ...",Tembara Coastal Cluster,2.20,104.55,2010-01-01,ambient;secure_cage;2-8C;frozen
4,FCL-1016,Northstar Tembara Cold-Chain Hub,warehouse,BP-30047,"Northstar Tembara Cold-Chain Hub, Tembara Coas...",Tembara Coastal Cluster,2.23,104.58,2008-01-01,ambient;secure_cage;2-8C;frozen


###2.8 Nettoyage — erp_business_partners
partner_role présente une incohérence de casse ("Supplier" / "supplier"), comme observé précédemment pour item_category dans purchase_order_lines. Elle est uniformisée en minuscules.

In [ ]:
df_bp['partner_role'] = df_bp['partner_role'].str.lower()
df_bp['partner_role'].value_counts()

,count
partner_role,
customer,32
service_partner,19
supplier,14
carrier,9
supplier_customer,6


external_registration_number (14 valeurs manquantes) et billing_region (1 valeur manquante) ne présentent pas de motif apparent expliquant leur absence. Ces cas sont rendus explicites avec la catégorie "non_renseigne", dans la continuité de la démarche adoptée pour les autres fichiers

In [ ]:
df_bp['external_registration_number'] = df_bp['external_registration_number'].fillna('non_renseigne')
df_bp['billing_region'] = df_bp['billing_region'].fillna('non_renseigne')

Point de vigilance pour l'analyse : la colonne shipping_region mélange deux granularités géographiques différentes (ex : "Valenne Central District" face à "Valenne Republic"). Cette hétérogénéité n'est pas corrigée ici mais devra être prise en compte lors d'agrégations géographiques (à faire lors de l'analyse)

In [ ]:
dfs["business_partners"] = df_bp
df_bp.head()

,partner_code,partner_name,partner_role,external_registration_number,billing_region,shipping_region,default_currency,payment_terms,partner_status,risk_class,onboarding_date
0,BP-30001,Marula Cold Store Cooperative,supplier_customer,REG-AVL-20012,Avelora Federation,Eastern Avelora Agricultural Cluster,AVL,NET45,active,review,2001-02-01
1,BP-30002,Blue Heron Monitoring Cooperative,customer,REG-KNT-20037,Kintara Republic,Kintara Karst Cluster,KRN,NET45,active,standard,2012-03-01
2,BP-30003,Orra Technical Services,supplier,REG-AVL-20022,Avelora Federation,Eastern Avelora Agricultural Cluster,AVL,PUBLIC30,active,review,2008-04-01
3,BP-30004,Common Ground Climate Forum,service_partner,REG-VLN-01008,Valenne Republic,Valenne Central District,VLN,PREPAID,active,review,2014-05-01
4,BP-30005,Warden Medical Ventures,service_partner,REG-VLN-01016,Valenne Republic,Valenne Republic,VLN,PREPAID,inactive,standard,2013-06-01


###2.9 Nettoyage — shipments

Les doublons et la colonne constante (shipment_leg) ont déjà été traités dans les étapes 2.1 et 2.2. Il reste des incohérences de casse sur deux colonnes catégorielles, un problème d'espaces multiples sur une troisième, et des valeurs manquantes sur arrival_time.

In [ ]:
df_shp['temperature_control_class'] = df_shp['temperature_control_class'].str.lower()
df_shp['shipment_status'] = df_shp['shipment_status'].str.lower()

df_shp['temperature_control_class'].value_counts()

,count
temperature_control_class,
ambient,343
2-8c,103
frozen,4


In [ ]:
df_shp['shipment_status'].value_counts()

,count
shipment_status,
delivered,385
delayed,34
in_transit,17
returned,14


declared_purpose présente un problème différent des cas précédents : certaines valeurs contiennent des espaces multiples ou en début de chaîne (ex : " laboratory delivery" au lieu de "laboratory delivery"), ce qui crée artificiellement de fausses catégories distinctes. Une fonction dédiée est utilisée pour normaliser les espaces, sur le même principe que nettoyer_colonne_numerique.

In [ ]:
def nettoyer_espaces(serie):
    s = serie.astype(str).str.strip()
    s = s.str.replace(r'\s+', ' ', regex=True)
    return s

df_shp['declared_purpose'] = nettoyer_espaces(df_shp['declared_purpose'])
df_shp['declared_purpose'].value_counts()

,count
declared_purpose,
scheduled stock transfer,74
laboratory delivery,72
field programme support,72
customer distribution,64
facility replenishment,64
agricultural distribution,60
temperature-controlled clinical programme delivery,26
environmental and wildlife reference materials,18


vérification, pas de modification

In [ ]:
df_shp[df_shp['arrival_time'].isna()]['shipment_status'].value_counts()

,count
shipment_status,
in_transit,17


mise à jour du dict

In [ ]:
dfs["shipments"] = df_shp
df_shp.head()

,shipment_number,dispatch_time,arrival_time,sender_partner_code,receiver_partner_code,origin_facility_code,destination_facility_code,carrier_partner_code,item_category,declared_quantity,declared_unit,temperature_control_class,authorization_class,declared_purpose,shipment_status
0,SHP-2014-00001,2014-07-10T09:00:00,2014-07-11T03:00:00,BP-30078,BP-30021,FCL-1011,FCL-1007,BP-30047,biological_sample,2,sealed_lot,2-8c,expedited,temperature-controlled clinical programme deli...,delivered
1,SHP-2015-00001,2015-01-10T11:00:00,2015-01-13T16:00:00,BP-30035,BP-30033,FCL-1009,FCL-1013,BP-30020,medical_consumables,73,case,2-8c,ordinary,scheduled stock transfer,returned
2,SHP-2015-00002,2015-01-11T06:00:00,2015-01-12T00:00:00,BP-30028,BP-30037,FCL-1020,FCL-1017,BP-30015,office_supply,82,case,ambient,emergency,customer distribution,delivered
3,SHP-2015-00003,2015-02-17T15:00:00,2015-02-19T08:00:00,BP-30056,BP-30078,FCL-1019,FCL-1011,BP-30047,biological_sample,2,sealed_lot,2-8c,restricted,laboratory delivery,delivered
4,SHP-2015-00004,2015-06-10T12:00:00,2015-06-12T09:00:00,BP-30025,BP-30079,FCL-1026,FCL-1022,BP-30025,legal_service,1,service,ambient,ordinary,facility replenishment,delivered


In [ ]:
#simple verif
for nom, df in dfs.items():
    print(nom, ":", df.shape)

facilities : (30, 10)
business_partners : (80, 11)
inventory_month_end : (500, 11)
inventory_adjustments : (140, 11)
goods_receipts : (610, 11)
shipments : (450, 15)
purchase_order_lines : (650, 15)


#### Bilan nettoyage — logistics

On a nettoyé les 7 fichiers un par un en suivant à chaque fois la même logique :

- on a supprimé les doublons (lignes identiques) et les colonnes qui ne varient jamais (ex : `facility_status`, `shipment_leg`), elles ne servent à rien pour l'analyse
- on a corrigé les colonnes numériques qui étaient stockées en texte (virgule au lieu de point, unité collée genre "12kg"), avec la fonction `nettoyer_colonne_numerique`
- on a uniformisé les catégories qui avaient des soucis de casse (ex : "Supplier" vs "supplier", "DELIVERED" vs "delivered"), en mettant tout en minuscules
- pour les valeurs manquantes, on a essayé à chaque fois de comprendre pourquoi elles étaient manquantes avant de faire quoi que ce soit. Quand on n'avait pas d'explication, on a préféré les marquer clairement comme `"non_renseigne"` plutôt que d'inventer une valeur. Et quand l'absence avait un sens (ex : pas de `arrival_time` parce que le colis est encore `in_transit`), on l'a laissée telle quelle

**Trucs à garder en tête pour la suite** :
- `unit_price` est dans 5 devises différentes, donc pas comparable tel quel entre commandes
- `shipping_region` mélange des noms de région et des noms de pays, à uniformiser avant de faire des stats dessus

**Ce qu'il reste à faire** : on n'a pas encore regardé les valeurs aberrantes (outliers) dans les colonnes numériques, c'est la prochaine étape avant de dire que le nettoyage logistics est vraiment terminé.

**Public records**

In [ ]:
df_pr = pd.read_csv("data/public_records/institutional_memos.csv")  #à finir quand on saura faire les textes

In [ ]:
print(df_pr.head())

# Valeurs manquantes
print(df_pr.isna().sum())

# Nombre de valeurs uniques
print(df_pr.nunique())

  document_id document_date          document_type     author_name  \
0  MEM-510062    2024-04-30  inventory_explanation      Soren Kade   
1  MEM-510100    2023-08-17   security_instruction      Soren Kade   
2  MEM-510101    2025-03-14         meeting_record      Soren Kade   
3  MEM-510128    2018-05-15    internal_memorandum  Adrian Vosmere   
4  MEM-510069    2021-10-22    internal_memorandum  Dr. Elias Voss   

     author_organization                              recipient_text  \
0  Meridian BioLogistics  Restricted operations and compliance group   
1  Meridian BioLogistics  Restricted operations and compliance group   
2  Meridian BioLogistics  Restricted operations and compliance group   
3    Vosmere Biotechnics         Operations, finance, and compliance   
4  Meridian BioLogistics         Operations, finance, and compliance   

                                  subject_line  \
0         April Biological-Unit Reconciliation   
1  Secure Delivery Record-Handling Instruction

In [ ]:
print(df_pr.head())

# Valeurs manquantes
print(df_pr.isna().sum())

# Nombre de valeurs uniques
print(df_pr.nunique())

  document_id document_date          document_type     author_name  \
0  MEM-510062    2024-04-30  inventory_explanation      Soren Kade   
1  MEM-510100    2023-08-17   security_instruction      Soren Kade   
2  MEM-510101    2025-03-14         meeting_record      Soren Kade   
3  MEM-510128    2018-05-15    internal_memorandum  Adrian Vosmere   
4  MEM-510069    2021-10-22    internal_memorandum  Dr. Elias Voss   

     author_organization                              recipient_text  \
0  Meridian BioLogistics  Restricted operations and compliance group   
1  Meridian BioLogistics  Restricted operations and compliance group   
2  Meridian BioLogistics  Restricted operations and compliance group   
3    Vosmere Biotechnics         Operations, finance, and compliance   
4  Meridian BioLogistics         Operations, finance, and compliance   

                                  subject_line  \
0         April Biological-Unit Reconciliation   
1  Secure Delivery Record-Handling Instruction

In [ ]:
constant_cols = [
    col for col in df.columns
    if df_pr[col].nunique() <= 1
]

df_pr = df_pr.drop(columns=constant_cols)

In [ ]:
for col in df_pr.select_dtypes(include="object").columns:
    print("\n---", col, "---")
    print("Nombre de catégories :", df_pr[col].nunique())
    print(df_pr[col].value_counts().head(10))


--- document_id ---
Nombre de catégories : 30
document_id
MEM-510062    1
MEM-510100    1
MEM-510101    1
MEM-510128    1
MEM-510069    1
MEM-510030    1
MEM-510080    1
MEM-510059    1
MEM-510074    1
MEM-510018    1
Name: count, dtype: int64

--- document_date ---
Nombre de catégories : 30
document_date
2024-04-30    1
2023-08-17    1
2025-03-14    1
2018-05-15    1
2021-10-22    1
2024-03-02    1
2019-08-09    1
2022-01-16    1
2025-06-23    1
2020-11-03    1
Name: count, dtype: int64

--- document_type ---
Nombre de catégories : 6
document_type
inventory_explanation    5
security_instruction     5
meeting_record           5
internal_memorandum      5
procurement_note         5
shipment_report          5
Name: count, dtype: int64

--- author_name ---
Nombre de catégories : 27
author_name
Soren Kade        4
Adrian Vosmere    1
Dr. Elias Voss    1
Jacob Vale        1
Amara Senn        1
Niko Verran       1
Celene Varro      1
Oskar Nivren      1
Adrienne Solva    1
Tomas Pell       

In [ ]:
# def extract_document_type(text):
#     text = str(text).strip().upper()

#     if text.startswith("SUBJECT"):
#         return "SUBJECT"
#     elif text.startswith("SCOPE"):
#         return "SCOPE"
#     elif text.startswith("MEETING RECORD"):
#         return "MEETING"
#     elif text.startswith("REQUEST"):
#         return "REQUEST"
#     elif text.startswith("ROUTE"):
#         return "ROUTE"
#     elif text.startswith("PERIOD"):
#         return "PERIOD"
#     else:
#         return "OTHER"


# df_pr["document_type"] = df["text"].apply(extract_document_type)
# df_pr = df_pr.drop('text', axis=1)
df_pr.to_csv("data/public_records/institutional_memos.csv", index=False)

licensed_facilities.csv

In [ ]:
df_lf = pd.read_csv("data/public_records/licensed_facilities.csv") #à finir quand on saura faire les textes

on enleve cette tab car elle est vide elle ne contient rien

In [ ]:
df_lf = df_lf.drop('license_end', axis=1)
df_lf = df_lf.drop('facility_status', axis=1)


In [ ]:
print(df_lf.head)

<bound method NDFrame.head of   facility_license_number                       facility_name  \
0             FAC-VLN-101    Asterion Central Clinical Campus   
1             FAC-AVL-201   Asterion Avelora Executive Clinic   
2             FAC-AVL-301  Threshold Earth Avelora Laboratory   

     facility_category operating_organization_registration  \
0    research_hospital                       REG-VLN-01002   
1       private_clinic                       REG-VLN-01002   
2  research_laboratory                       REG-AVL-01007   

                                      street_address              region  \
0  Asterion Central Clinical Campus, Valenne Cent...    Valenne Republic   
1  Asterion Avelora Executive Clinic, Eastern Ave...  Avelora Federation   
2  Threshold Earth Avelora Laboratory, Eastern Av...  Avelora Federation   

   latitude  longitude license_start  \
0     47.20       7.51    2001-02-01   
1      7.35     101.62    2008-05-15   
2      7.50     101.75    2012-01-1

licensed_professionals.csv

In [ ]:
df_lp = pd.read_csv("data/public_records/licensed_professionals.csv")

In [ ]:
# Valeurs manquantes
# print(df_lp.isna().sum())
# print(len(df_lp))

# Nombre de valeurs uniques
print(df_lp.nunique())

license_number              53
professional_name           53
profession                   5
specialty                   16
license_issue_date          40
license_expiry_date          0
license_status               1
primary_facility_license     3
registered_region            5
disciplinary_status          2
dtype: int64


In [ ]:
print(df_lp.columns)

Index(['license_number', 'professional_name', 'profession', 'specialty',
       'license_issue_date', 'license_expiry_date', 'license_status',
       'primary_facility_license', 'registered_region', 'disciplinary_status'],
      dtype='object')


on a décidé de les supprimer car ils ont tous la même valeur ou sont vides

In [ ]:
df_lp = df_lp.drop('license_status', axis=1)
df_lp = df_lp.drop('license_expiry_date', axis=1)

In [ ]:
df_lp['license_issue_date'] = pd.to_datetime(df_lp['license_issue_date'], format='%Y-%m-%d', errors='coerce')
df_lp[df_lp['license_issue_date'].isna()]
df_lp.loc[26, 'license_issue_date'] = pd.to_datetime('1994-07-01', format='%Y-%m-%d')

In [ ]:
print(df_lp.iloc[26])

license_number                                LIC-TMB-ADM-1027
professional_name                                  Doran Maren
profession                  public administration professional
specialty                             programme administration
license_issue_date                         1994-07-01 00:00:00
primary_facility_license                                   NaN
registered_region                                     Tembaran
disciplinary_status                                       none
Name: 26, dtype: object
